type-metadata/src/lib.rs

```diff
@@ -1,7 +1,319 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		assert_eq!(2 + 2, 4);
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TypeIdent {
+	pub namespace: Vec<String>,
+	pub ident: IdentKind,
+	pub args: Vec<TypeIdent>,
+}
+
+impl TypeIdent {
+	pub fn new(ident: IdentKind) -> TypeIdent {
+		TypeIdent {
+			namespace: vec![],
+			ident,
+			args: vec![],
+		}
+	}
+	pub fn namespace(mut self, namespace: Vec<String>) -> Self {
+		self.namespace = namespace;
+		self
+	}
+	pub fn args(mut self, args: Vec<TypeIdent>) -> Self {
+		self.args = args;
+		self
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(String),
+
+	// primitives, including build-in and common core/std primitives
+	Bool,
+	Str,
+	Unit,
+	U8,
+	U16,
+	U32,
+	Usize,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	Isize,
+	I64,
+	I128,
+	Array(u16),
+	Vector,
+	Tuple,
+	Option,
+	Result,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum TypeDef {
+	Primitive,
+	Struct(StructDef),
+	EnumDef(EnumDef),
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum FieldName {
+	Unnamed(u16),
+	Named(String),
+}
+#[derive(PartialEq, Eq, Debug)]
+pub struct Field {
+	pub name: FieldName,
+	pub ident: TypeIdent,
+}
+pub type StructDef = Vec<Field>;
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct EnumVariant {
+	pub name: String,
+	pub index: u16,
+	pub fields: Vec<Field>,
```

Comment:
I would again go for a different approach where we make a distintion between C-like enums and Rust-like enums (with data). They are quite different in many aspects. For example they can have different attributes and the former can also specify default values for its variants.